proof/invariant-abstract/DetSchedSchedule_AI.thy

```diff
@@ -907,23 +907,19 @@ lemma as_user_valid_sched[wp]:
                          st_tcb_def2 valid_blocked_def)
   done
 
-lemma switch_to_thread_ct_not_queued[wp]:
-  "\<lbrace>valid_queues\<rbrace> switch_to_thread t \<lbrace>\<lambda>rv s. not_queued (cur_thread s) s\<rbrace>"
-  apply (simp add: switch_to_thread_def)
-  including no_pre
-  apply wp
-     prefer 4
-     apply (rule get_wp)
-    prefer 3
-    apply (rule assert_inv)
-   prefer 2
-   apply (rule arch_switch_to_thread_valid_queues')
-  apply (simp add: tcb_sched_action_def
-                   tcb_sched_dequeue_def | wp)+
+lemma tcb_sched_action_dequeue_not_queued[wp]:
+  "\<lbrace>valid_queues\<rbrace> tcb_sched_action tcb_sched_dequeue t \<lbrace>\<lambda>_. not_queued t\<rbrace>"
+  unfolding tcb_sched_action_def tcb_sched_dequeue_def
+  apply wpsimp
   apply (clarsimp simp add: valid_queues_def etcb_at_def not_queued_def
                        split: option.splits)
   done
 
+lemma switch_to_thread_ct_not_queued[wp]:
+  "\<lbrace>valid_queues\<rbrace> switch_to_thread t \<lbrace>\<lambda>rv s. not_queued (cur_thread s) s\<rbrace>"
+  unfolding switch_to_thread_def
+  by wpsimp
```

Comment:
haha wow

Reply:
Yeah, this was one of the early cases before I found `classic_wp_pre` and when I was trying to properly fix broken proofs. The previous proof did a lot of very convoluted work to get to the `tcb_sched_action tcb_sched_dequeue` step with the appropriate precondition in place. Pulling it out to a separate lemma made things a lot simpler.